### ISSUE TITLE
Second fail in promise's then causes test run to stop

### ISSUE THREAD

**Issue by JCesarAB** [angular/protractor#2100]
Something very similar to Issue #1125 is happened to me. 

Any error in promise's then stops the execution and exit with message: "Process exited with error code 1".

Not only when the fail happen in an expect condition but also any error as for example looking an element. 

I'm using protractor 2.0.0 and cucumber 0.4.8

**Comment by JCesarAB** [comment]
The execution stops at second fail.

**Comment by trajic** [comment]
Exactly the same issue is happening to me.  On the second error in a promise the tests stop with Error Code 1.  I have the same setup at the user above (protractor 2.0.0 and cucumber 0.4.8).

**Comment by trajic** [comment]
Just FYI, I downgraded protractor to 1.8.0 and the scenarios continue past the 2nd failure (set of 10 scenarios that each fail with NoSuchElementError)

**Comment by sjelin** [comment]
Can you provide me with a minimal example?

**Comment by JCesarAB** [comment]
For sure, I try to create a simple example and I send it to you. 

Anyway, I have seen that the error has been introduced in the promise.js of the webdriver (protractor\node_modules\selenium-webdriver\lib\webdriver\promise.js) that is contained by the new version of selenium-webdriver (2.45.1) which comes with protractor 2.0.0.

I paste the stacktrace so that you can analyzing it:

      Error: Index out of bound. Trying to access element at index: -1, but there are only 0 elements that match locator by.repeater("none")
	at C:\Users\Zesar\workspace\cosmo-front-client\node_modules\grunt-protractor-runner\node_modules\protractor\lib\element.js:257:15
        at [object Object].promise.ControlFlow.runInFrame_ (C:\Users\Zesar\workspace\cosmo-front-client\node_modules\grunt-protractor-runner\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\promise.js:1877:20)
        at [object Object].promise.Callback_.goog.defineClass.notify (C:\Users\Zesar\workspace\cosmo-front-client\node_modules\grunt-protractor-runner\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\promise.js:2464:25)
        at [object Object].promise.Promise.notify_ (C:\Users\Zesar\workspace\cosmo-front-client\node_modules\grunt-protractor-runner\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\promise.js:563:12)
        at Array.forEach (native)
        at Object.goog.array.forEach (C:\Users\Zesar\workspace\cosmo-front-client\node_modules\grunt-protractor-runner\node_modules\protractor\node_modules\selenium-webdriver\lib\goog\array\array.js:203:43)
        at [object Object].promise.Promise.notifyAll_ (C:\Users\Zesar\workspace\cosmo-front-client\node_modules\grunt-protractor-runner\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\promise.js:552:16)
        at goog.async.run.processWorkQueue (C:\Users\Zesar\workspace\cosmo-front-client\node_modules\grunt-protractor-runner\node_modules\protractor\node_modules\selenium-webdriver\lib\goog\async\run.js:125:21)
        at runMicrotasksCallback (node.js:337:7)
        at process._tickDomainCallback (node.js:381:11)

**Comment by JCesarAB** [comment]
Hello, here is the example: https://github.com/JCesarAB/2100ProtractorIssueExample

**Comment by djlaplante** [comment]
I am also getting this error

**Comment by jlin412** [comment]
It also happens to me as well.  However, the case of is if I have multiple promises in a step and it fails in a non return promise and it will exit when I call the step the second time in a different scenario.
```
  this.When(/^I failed at the first promise when there are two protractor function calls$/, function () {
    $('bad_css_path').isDisplayed();
    return $('bab_css_path').isPresent();
  });
```

**Comment by jlin412** [comment]
Further investigation.  It could be a problem in cucumber.js how it handle exception for node 0.12.x.
https://github.com/cucumber/cucumber-js/issues/394

**Comment by juliemr** [comment]
Looks like a domains error with new version of node and cucumber, based on the issue linked above. I'm marking this as an external bug.

**Comment by jlin412** [comment]
@juliemr https://github.com/angular/protractor/pull/2615 this PR should fix this bug and please merge in the next build

**Comment by hankduan** [comment]
https://github.com/angular/protractor/pull/2615 is merged, and will be released with Protractor 3.0. Please confirm the fix once it's out.

**Comment by aluzardo** [comment]
Similar issue happening with protractor 4 https://github.com/angular/protractor/issues/3491 https://github.com/angular/protractor/issues/3553
